server/src/main/java/com/thoughtworks/go/server/service/ElasticAgentPluginService.java

```diff
@@ -308,12 +323,73 @@ public void jobCompleted(JobInstance job) {
+            String description = format("The job completion call to the plugin for the job identifier [%s] failed for secrets resolution: %s ", jobIdentifier.toString(), e.getMessage());
+            ServerHealthState healthState = error("Failed to notify plugin", description, general(scopeForJob(jobIdentifier)));
+            healthState.setTimeout(Timeout.FIVE_MINUTES);
+            serverHealthService.update(healthState);
+            LOGGER.error(description);
+        }
+    }
+
+    private void logToJobConsole(JobIdentifier identifier, String message) {
+        try {
+            consoleService.appendToConsoleLog(identifier, message);
+        } catch (IllegalArtifactLocationException | IOException e) {
+            LOGGER.error(format("Failed to add message(%s) to the job(%s) console", message, identifier), e);
+        }
+    }
+
+    private Predicate<JobPlan> isElasticAgent() {
+        return JobPlan::requiresElasticAgent;
+    }
+
+    private HealthStateScope scope(String pluginId) {
+        return HealthStateScope.aboutPlugin(pluginId, "missingPlugin");
+    }
+
+    @NotNull
+    private HealthStateScope scopeForJob(JobIdentifier jobIdentifier) {
+        return forJob(jobIdentifier.getPipelineName(), jobIdentifier.getStageName(), jobIdentifier.getBuildName());
+    }
+
+    private boolean resolveSecrets(String pluginId, List<ClusterProfile> clusterProfiles) {
+        try {
+            for (ClusterProfile clusterProfile : clusterProfiles) {
+                secretParamResolver.resolve(clusterProfile);
+            }
+        } catch (RulesViolationException | SecretResolutionFailureException e) {
+            String description = format("Secrets resolution failed for cluster profile associated with plugin [%s]: %s ", pluginId, e.getMessage());
+            serverHealthService.update(error(format("Ping failed for '%s' plugin", pluginId), description, general(scope(pluginId))));
+            LOGGER.error(description);
+            return false;
+        }
+        return true;
+    }
 
-        elasticAgentPluginRegistry.reportJobCompletion(pluginId, elasticAgentId, job.getIdentifier(), elasticProfileConfiguration, clusterProfileConfiguration);
+    private boolean resolveSecrets(JobIdentifier jobIdentifier, ClusterProfile clusterProfile, ElasticProfile elasticProfile) {
+        try {
+            if (clusterProfile != null)
+                secretParamResolver.resolve(clusterProfile);
+            secretParamResolver.resolve(elasticProfile);
+        } catch (RulesViolationException | SecretResolutionFailureException e) {
+            String cancellationMessage = format("\nThis job was cancelled by GoCD. This job is configured to run on an elastic agent, but the associated elastic configurations failed for secrets resolution: %s", e.getMessage());
+            logToJobConsole(jobIdentifier, cancellationMessage);
+            scheduleService.cancelJob(jobIdentifier);
```

Comment:
> You will also need to add a message on the JobStatusTopic so that the JobStatusListener is called.

This is taken care by the schedule service class.